_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+[Elasticsearch's Built-in Analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-analyzers.html).
+
+3. There is nothing special about `@Field` annotation which is applied to `hashTags` field. I just want to highlight the fact that tokenizer splits the text on any nonalphanumeric character. So a value like this '#food#health' will be 
+automatically tokenized to ['food','health].
+
+4. `tag` field which is an enum will mostly consist of a single word. We don't need to analyze such fields. So, instead, we can either set `analyze` property of `@Field` to 
+`Analyze.NO` or we can apply a `normalizer`. Hibernate will then treat this field as `keyword`. The lowercase normalizer
+that I have used here will be applied both at the time of indexing and searching. So, both 'MOVIE' or 'movie' value will be a match.  
+
+5. When we apply `@SortableField` over a field, Elasticsearch will optimize the index for sorting operations over those fields.
+We can still perform sorting operations over other fields which are not marked with this annotation but that will have some
+performance penalties.
+
+6. `@IndexedEmbedded` is used when we want to perform a search over nested objects fields. For instance, let's say we want to search
+all posts made by a user with the first name 'Joe' (user.first: joe). 
+
+
+Some things to note about analyzers and normalizers when using Elasticsearch integration:
+* Although you can use Lucene's built-in analyzers or even custom analyzers using `@AnalyzerDef`, it's recommended to use 
+built-in Elasticsearch analyzers.
+* Built-in Normalizers are directly translated to Elasticsearch supported normalizers but for this to work you 
+need to define Normalizer using `@NormalizerDef` and pass the name in definition as I have done in `Tag` enum's field.
+
+We need to cover a few more properties of `@Field` annotation before moving on. Let's look at `User` class for that.
+
+```java
+@Getter
+@Setter
+@Entity
+@Indexed(index = "idx_user")
+class User {
+  @Id
+  private String id;
+
+  @Field(store = Store.YES)//->[3]
+  @Field(name = "fullName")//->[2]
+  private String first;
+
+  @Field(index = Index.NO) //->[1]
+  private String middle;
+
+  @Field
+  @Field(name = "fullName")//->[2]
+  private String last;
+
+  @Field
+  private Integer age;
+
+  @ContainedIn//->[4]
+  @OneToMany(mappedBy = "user")
+  private List<Post> post;
+}
+```
+
+Couple of interesting things to notice here. Let's start with the `index` property.
+
+1. `Index.NO` this indicates that property will be stored in the index document, but it won't be indexed. We won't be able
+to perform any search operation over it. You might be thinking "Why not simply remove `@Field` annotation?". And the answer is that we still need this field for Projection (Covered in the next section).
```

Comment:
Instead of saying "Covered in the next section", do this:

And the answer is that we still need this field for [Projection](Link to the section that explains projection)